redirectpolicy: Add option to limit allowed addresses in AddressMatcher

joamaki opened the pull request:
This adds the `--lrp-address-matcher-cidrs` flag to limit which IP addresses can be used in an AddressMatcher. The corresponding helm option is `localRedirectPolicies.addressMatcherCIDRs`.

The `localRedirectPolicy` helm option is marked deprecated and `localRedirectPolicies.enabled` is introduced instead to unify LRP related configuration under one prefix.

If an LRP is created with an AddressMatcher containing a disallowed IP address then a log message is emitted ("Ignoring IP in AddressMatcher") and the LRP parses to an empty one with type "none" which will undo any prior reconciled configuration.

```release-note
LocalRedirectPolicy: Helm option 'localRedirectPolicy' is now deprecated and 'localRedirectPolicies.enabled' should be used instead. The new helm option localRedirectPolicies.addressMatcherCIDRs can be used to limit which addresses are allowed in an AddressMatcher.
```

joamaki committed:
redirectpolicy: Add lrp-address-matcher-cidrs option

Add support for limiting which IP addresses are allowed to be used with the
AddressMatcher rule.

E.g. if --lrp-address-matcher-cidrs=10.0.0.0/8 then the
[CiliumLocalRedirectPolicy.Spec.RedirectFrontend.AddressMatcher.IP] must
be within the CIDR 10.0.0.0/8. If it is not, then the LRP will be parsed as
type "none" with no frontends or backends and removing any prior configuration.

A warning log message "Ignoring IP in AddressMatcher" is emitted when an address
does not match given cidrs.

By default --lrp-address-matcher-cidrs default to empty set of cidrs in which case
any address is allowed.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
helm: Add localRedirectPolicies.addressMatcherCIDRs option

Add a helm option for the --lrp-address-matcher-cidrs flag.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

gandro commented on install/kubernetes/cilium/values.yaml:
```diff
@@ -2121,6 +2121,13 @@ l2NeighDiscovery:
 l7Proxy: true
 # -- Enable Local Redirect Policy.
 localRedirectPolicy: false
+localRedirectPolicies:
```
If we introduce a new section, I think we could introduce `localRedirectPolicies.enabled` as an alias for `localRedirectPolicy` and deprecate the latter. Here's an example of how this can be done: https://github.com/cilium/cilium/pull/15846

joamaki replied:
Seems reasonable. Let's go with that.

gandro commented on install/kubernetes/cilium/templates/cilium-configmap.yaml:
```diff
@@ -731,6 +731,12 @@ data:
 {{- end }}
 
+{{- if hasKey .Values "localRedirectPolicies" }}
+{{- if hasKey .Values.localRedirectPolicies "addressMatcherCIDRs" }}
```
This field is `nil` by default, so this will always evaluate to true I believe. Why not use a truthiness check, i.e. `if .Values.localRedirectPolicies.addressMatcherCIDRs`?

joamaki replied:
That seems to work, thanks!

gandro reviewed (commented):
Thanks! Some minor feedback regarding Helm

networkop reviewed (approved):
lgtm

aditighag commented on pkg/loadbalancer/redirectpolicy/redirectpolicy.go:
```diff
@@ -218,6 +211,25 @@ func getSanitizedLocalRedirectPolicy(name, namespace string, uid types.UID, spec
+				logfields.K8sNamespace, namespace,
+				logfields.Name, name)
+
+			// Return an "empty" policy so that any prior policy will be flushed out.
```
Why do we clear out prior state -- is it not enough to just reject the current policy that specifies an address disallowed by `AddressMatcherCIDRS`?

joamaki replied:
Yeah good point. It would retain the old policy, but that's likely what most users would expect anyway. Will change it to reject instead.

joamaki replied:
Ok, so when testing this I noticed we'll get into trouble if we just reject it when parsing. If we delete a policy that passed and then add quickly a policy that doesn't then we'll end up never processing the deletion and end up keeping an old "deleted" policy. 

Fixed this by always trying to delete policy if parsing fails to be safe against these sort of situations.

aditighag replied:
I don't quite follow -- if a frontend address specified in an LRP is disallowed, then we reject the policy. The LRP manager plumbs a redirect service entry for a frontend -> backend(s) as derived from an LRP.  What prior state are you referring to -- it is some state stored in stateDB or the BPF LB maps?

Maybe if you give an example, that'll help me understand the scenario you you are highlighting better.

joamaki replied:
The main issue here is that we are not guaranteed to see a deletion event if there's a deletion followed by creation for the object with the same namespace+name. This is due to how the Indexer and Resource[T] process the deltas: the creation replaces the deletion if it happens quickly.

In tests it's not hard to be fast enough to cause the deletion to be replaced by the subsequent creation and if we just "reject" the creation while parsing we'll end up keeping the `LocalRedirectPolicy` object that the user had deleted earlier. That to me seems wrong. Even if this situation is very hard to trigger in production I'd much rather be safe than sorry.

Of course this has the flip side that if the user edits the policy and it has an issue in it we'll just end up deleting whatever prior policy was in place. With LRP this is much less of an issue though since the policies are mostly immutable (though the reason why they were made immutable has been fixed...).

If you're happy with this explanation please mark this resolved so that the PR gets auto-merged.

aditighag replied:
>Of course this has the flip side that if the user edits the policy and it has an issue in it we'll just end up deleting whatever prior policy was in place. 

This makes me nervous. 
But, as you said, LRPs are immutable. So any updates to an existing LRP will get rejected when users try to apply the LRP, and hopefully we'll never hit block of code. /cc @ysksuzuki

aditighag reviewed (commented):
Posted one clarifying question. 
- What about restoring previously plumbed LRP entries that may violate user-configured address matcher cidrs?
- Also, how do you feel about documenting this flag - https://docs.cilium.io/en/latest/network/kubernetes/local-redirect-policy/? 
- The flag by default allows all CIDRs, so I suppose there is no upgrade impact after this changre?

tommyp1ckles commented on Documentation/helm-values.rst:
```diff
@@ -2732,8 +2732,16 @@
      - List of ports from service to be automatically redirected to above backend. Any service exposing one of these ports will be automatically redirected. Fine-grained control can be achieved by using the service annotation.
```
There's also references to the now deprecated `localRedirectPolicy` value in network/kubernetes/local-redirect-policy.rst that we may want to update.

joamaki replied:
Updated.

joamaki committed:
Documentatation: Update local redirect policies with new options

Added a section on the 'localRedirectPolicies.addressMatcherCIDRs' and
changed 'localRedirectPolicy' to 'localRedirectPolicies.enabled'.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

aditighag reviewed (approved):
> @cilium/sig-lb looks like this PR is awaiting your review, could you take a look?

There was an unresolved comment.
